src/WebJobs.Script/Config/HostJsonFileConfigurationSource.cs

```diff
@@ -125,68 +129,77 @@ private void ProcessArray(JArray jArray)
-                // The "startup" logger is used in this class for startup related logs. The public logger is used
-                // for all other logging after startup.
-
-                ScriptApplicationHostOptions options = _configurationSource.HostOptions;
-                string hostFilePath = Path.Combine(options.ScriptPath, ScriptConstants.HostMetadataFileName);
-                JObject hostConfigObject = LoadHostConfig(hostFilePath);
-                InitializeHostConfig(hostFilePath, hostConfigObject);
-                string sanitizedJson = SanitizeHostJson(hostConfigObject);
-
-                _logger.HostConfigApplied();
-
-                // Do not log these until after all the configuration is done so the proper filters are applied.
-                _logger.HostConfigReading(hostFilePath);
-                _logger.HostConfigRead(sanitizedJson);
-
-                return hostConfigObject;
+                using (_metricsLogger.LatencyEvent(MetricEventNames.LoadHostConfigurationSource))
+                {
+                    // Before configuration has been fully read, configure a default logger factory
+                    // to ensure we can log any configuration errors. There's no filters at this point,
+                    // but that's okay since we can't build filters until we apply configuration below.
+                    // We'll recreate the loggers after config is read. We initialize the public logger
+                    // to the startup logger until we've read configuration settings and can create the real logger.
+                    // The "startup" logger is used in this class for startup related logs. The public logger is used
+                    // for all other logging after startup.
+
+                    ScriptApplicationHostOptions options = _configurationSource.HostOptions;
+                    string hostFilePath = Path.Combine(options.ScriptPath, ScriptConstants.HostMetadataFileName);
+                    JObject hostConfigObject = LoadHostConfig(hostFilePath);
+                    InitializeHostConfig(hostFilePath, hostConfigObject);
+                    string sanitizedJson = SanitizeHostJson(hostConfigObject);
+
+                    _logger.HostConfigApplied();
+
+                    // Do not log these until after all the configuration is done so the proper filters are applied.
+                    _logger.HostConfigReading(hostFilePath);
+                    _logger.HostConfigRead(sanitizedJson);
+
+                    return hostConfigObject;
+                }
             }
 
             private void InitializeHostConfig(string hostJsonPath, JObject hostConfigObject)
             {
-                // If the object is empty, initialize it to include the version and write the file.
-                if (!hostConfigObject.HasValues)
+                using (_metricsLogger.LatencyEvent(MetricEventNames.InitializeHostConfiguration))
```

Comment:
is this needed? metrics for LoadHostConfigurationFile should be enough

Reply:
This does a TryWriteHostJson which is an i/o operation. Thought it might be useful to have this in addition to LoadHostConfigurationSource to know the difference in total time against i/o time 

---
In reply to: [333798554](https://github.com/Azure/azure-functions-host/pull/4973#discussion_r333798554) [](ancestors = 333798554)